### ISSUE TITLE
RFC: using external project types

### ISSUE THREAD

**Issue by mbonig** [eladb/projen#193]
It's unclear what the steps and process looks like to use projen with external project types. There is a chicken&egg problem with getting the project type installed and usable.

**Comment by shortjared** [comment]
`~/.projenrc` global lookup so `npx projen discover` is a one time thing perhaps?

**Comment by mbonig** [comment]
I like the discovery aspect to it. I think we could reduce it to one step. e.g.:

```
npm projen new --global npm-module-name
```

telling the cli that instead of looking at the local set of projects shipped with the cli, it goes instead to the 'global' (npm) to get the project. I need to read more into the inventory class in the existing code, but I suspect changes would be around there to support this 'global' inventory. There'd have to be a strong opinion about the how the project was exposed to the `inventory` routine.

**Comment by shortjared** [comment]
This does bring in to question the concept of version pinning / updates. Can I say "always just use the latest" from the npm repository.

Also, we are looking to do this with strictly private packages.

**Comment by kcwinner** [comment]
I edited my comment to be what I originally meant. Discover via module, then create a project type from the included module. 
This would allow for naming overlap of project types but avoid npm module conflicts. And, if desired, the potential to have multiple project types registered via one module/mono-repo.

```
npx projen discover --global my-company-module
npx projen new --global my-company-project-ts
```

But I also like the simplicity of one package equals one project and one command to generate.

**Comment by kcwinner** [comment]
Ok so I've been experimenting with a `discover` command this morning. Since projen automatically discovers new types based on the `.jsii` output - what if we grab the module, extract it, read the `.jsii` file for types, copy and rename the source files (so as not to overlap any `index.ts`), inject into the local projen `.jsii` types. Also needs to add the module to project dependencies.

[Example discover command gist](https://gist.github.com/kcwinner/5962df02a43e4f46955038d9a60f67a8) 

This method leaves all projen new inventory *the same* and only adds a new discover command for adding project types. This also would require that the projects are of type `JsiiProject` (if using projen). Example external output will looks something like:
```
"types": {
    "cdk-appsync-project.AwsCdkAppSyncApp": {
      "assembly": "cdk-appsync-project",
      "base": "projen.AwsCdkTypeScriptApp",
      "docs": {
        "custom": {
          "pjid": "awscdk-appsync-app-ts"
        },
        "stability": "experimental",
        "summary": "AWS CDK AppSync Transformer App in TypeScript."
      },
      "fqn": "cdk-appsync-project.AwsCdkAppSyncApp",
      "initializer": {
        "docs": {
          "stability": "experimental"
        },
        "locationInModule": {
          "filename": "src/index.ts",
          "line": 20
        },
        "parameters": [
          {
            "name": "options",
            "type": {
              "fqn": "cdk-appsync-project.AwsCdkAppSyncAppOptions"
            }
          }
        ]
      },
      "kind": "class",
      "locationInModule": {
        "filename": "src/index.ts",
        "line": 19
      },
      "name": "AwsCdkAppSyncApp"
    },
    "cdk-appsync-project.AwsCdkAppSyncAppOptions": {
      "assembly": "cdk-appsync-project",
      "datatype": true,
      "docs": {
        "stability": "experimental"
      },
      "fqn": "cdk-appsync-project.AwsCdkAppSyncAppOptions",
      "interfaces": [
        "projen.AwsCdkTypeScriptAppOptions"
      ],
      "kind": "interface",
      "locationInModule": {
        "filename": "src/index.ts",
        "line": 5
      },
      "name": "AwsCdkAppSyncAppOptions",
      "properties": [
        {
          "abstract": true,
          "docs": {
            "default": "v1.63.0-rc.2",
            "stability": "experimental",
            "summary": "cdk-appsync-transformer version to use."
          },
          "immutable": true,
          "locationInModule": {
            "filename": "src/index.ts",
            "line": 11
          },
          "name": "transformerVersion",
          "type": {
            "primitive": "string"
          }
        }
      ]
    }
  }
```

I haven't tested with npx. Not sure if it will overwrite the `.jsii` file and require a new discover.

**Comment by mbonig** [comment]
Sounds very promising!

**Comment by kcwinner** [comment]
Here is what I currently have: [compare](https://github.com/eladb/projen/compare/master...kcwinner:feat/external-project-discovery). I would submit it as a PR but I think it needs further discussion to see if this behavior is desired. 

The process currently grabs the module using `npm pack`, extracts contents, finds the `.jsii` for project type discovery (requires that it's a projen type), checks to make sure there is a base of `projen.Project`, then adds the assembly to projen as well as the types to the projen jsii.

With this method a new external project looks like
```
npm init -y
npm install projen # Have to install so npx will use the local copy instead of child process
npx projen discover NPM-MODULE # Discover your npm module. Can have multiple project types in case you want to maintain them all together
npx projen new my-cool-project # Create the new project and synth

# This last step may not be needed but since projen reinstalls projen over my local copy it is 
npx projen discover NPM-module
```

For any CI system you would have to discover your project each time
```
yarn install
npx projen discover
npx projen
...
```

Pros:
* JSII discovery
* Works with current inventory system
* Should work with versioning, but currently doesn't bring the module in as a dependency
  * Example: `npx project discover mymodule@1.0.1`

Cons:
* Have to discover on each projen upgrade
* Not quite a one liner for creation
  * Could be modified to be something like `npx projen new my-cool-project --npm-module NPM-MODULE`

Any other thoughts @eladb ?

**Comment by eladb** [comment]
@kcwinner do we really need `discover`?

Here is what I've been thinking from a developer experience perspective.

## Projects in external modules

_projen_ is bundled with many project types out of the box, but it can also work with project types and components defined in external jsii modules (the reason we need jsii is because projen uses the jsii metadata to discover project types & options in `projen new`).

Say we have a module in npm called `projen-vue` which includes a single project type for [vue.js](https://vuejs.org/):

```shell
$ npx projen new --from projen-vue
```

Under the hood, `projen new` will install the `projen-vuejs` module from npm, discover that it contains a single project type and bootstrap it. If you examine your `.projenrc.js` file, you'll see that `projen-vuejs` is defined as a build dependency:

```js
const { VueJsProject } = require('projen-vuejs');

const project = new VueJsProject({
  name: 'my-vuejs-sample',
  // ...
  buildDeps: [
    'projen-vuejs'
  ]
});

project.synth();
```

> You can use any package syntax supported by [yarn add](https://classic.yarnpkg.com/en/docs/cli/add#toc-adding-dependencies) like `projen-vuejs@1.2.3`, `file:/path/to/local/folder`, `git@github.com/awesome/projen-vuejs#1.2.3`, etc.

**Comment by kcwinner** [comment]
@eladb I added discover more as a way to not interfere with the current system. Spoke about it with mbonig in the slack a bit yesterday and landed on removing the discover.

I'll modify it to fit the above this week. It should be easy enough given that discover & new combination already works.

Should it also support other package managers (pypi, maven, etc). Or should the assumption be you wrote it using a jsii project so it's already in TS and bundled to NPM by default?

**Comment by eladb** [comment]
Resolved by #212
